From 43ab7290012cdfb7a7fe28519cc70fd82a28a8ff Mon Sep 17 00:00:00 2001
From: Lars-Peter Clausen <lars@metafoo.de>
Date: Tue, 12 Feb 2013 17:14:01 +0100
Subject: [PATCH 225/509] DRM: axi-hdmi: Add support for RGB output

https://github.com/analogdevicesinc/linux.git xcomm_zynq
commit 10d81fe07f81c4ca5003c0c2e4713b4e6dee50a2

Some versions of the core output a RGB instead of a YCbCr signal. This patch
adds support for them.

Signed-off-by: Lars-Peter Clausen <lars@metafoo.de>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.c     |    2 ++
 drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.h     |    1 +
 drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_encoder.c |   22 ++++++++++++++--------
 3 files changed, 17 insertions(+), 8 deletions(-)

diff --git a/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.c b/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.c
index b75396c..2c74f00 100644
--- a/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.c
+++ b/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.c
@@ -169,6 +169,8 @@ static int axi_hdmi_platform_probe(struct platform_device *pdev)
 	slave_node = of_parse_phandle(pdev->dev.of_node, "encoder-slave", 0);
 	if (!slave_node)
 		return -EINVAL;
+
+	private->is_rgb = of_property_read_bool(pdev->dev.of_node, "adi,is-rgb");
 	
 	private->encoder_slave = of_find_i2c_device_by_node(slave_node);
 	of_node_put(slave_node);
diff --git a/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.h b/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.h
index 2b8ab90..5eb5d49 100644
--- a/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.h
+++ b/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_drv.h
@@ -33,6 +33,7 @@ struct axi_hdmi_private {
 	struct clk *hdmi_clock;
 
 	struct xlnx_pcm_dma_params dma_params;
+	bool is_rgb;
 };
 
 #endif
diff --git a/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_encoder.c b/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_encoder.c
index 754679c..37a3d24 100644
--- a/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_encoder.c
+++ b/drivers/gpu/drm/adi_axi_hdmi/axi_hdmi_encoder.c
@@ -264,16 +264,22 @@ static void axi_hdmi_encoder_dpms(struct drm_encoder *encoder, int mode)
 		hdmi_avi_infoframe_init(&config.avi_infoframe);
 
 		config.avi_infoframe.scan_mode = HDMI_SCAN_MODE_UNDERSCAN;
-		config.csc_scaling_factor = ADV7511_CSC_SCALING_4;
-		config.csc_coefficents = adv7511_csc_ycbcr_to_rgb;
 
-		if ((connector->display_info.color_formats & DRM_COLOR_FORMAT_YCRCB422) &&
-			config.hdmi_mode) {
-			config.csc_enable = false;
-			config.avi_infoframe.colorspace = HDMI_COLORSPACE_YUV422;
+		if (private->is_rgb) {
+				config.csc_enable = false;
+				config.avi_infoframe.colorspace = HDMI_COLORSPACE_RGB;
 		} else {
-			config.csc_enable = true;
-			config.avi_infoframe.colorspace = HDMI_COLORSPACE_RGB;
+			config.csc_scaling_factor = ADV7511_CSC_SCALING_4;
+			config.csc_coefficents = adv7511_csc_ycbcr_to_rgb;
+
+			if ((connector->display_info.color_formats & DRM_COLOR_FORMAT_YCRCB422) &&
+				config.hdmi_mode) {
+				config.csc_enable = false;
+				config.avi_infoframe.colorspace = HDMI_COLORSPACE_YUV422;
+			} else {
+				config.csc_enable = true;
+				config.avi_infoframe.colorspace = HDMI_COLORSPACE_RGB;
+			}
 		}
 
 		sfuncs->set_config(encoder, &config);
-- 
1.7.5.4

